src/main/java/io/airlift/units/DataSize.java

```diff
@@ -131,27 +194,37 @@ public static DataSize valueOf(String size)
         requireNonNull(size, "size is null");
         checkArgument(!size.isEmpty(), "size is empty");
 
-        Matcher matcher = PATTERN.matcher(size);
-        if (!matcher.matches()) {
+        Matcher longOrDouble = DECIMAL_WITH_UNIT_PATTERN.matcher(size);
+        if (!longOrDouble.matches()) {
             throw new IllegalArgumentException("size is not a valid data size string: " + size);
         }
-
-        double value = Double.parseDouble(matcher.group(1));
-        String unitString = matcher.group(2);
-
-        for (Unit unit : Unit.values()) {
-            if (unit.getUnitString().equals(unitString)) {
-                return new DataSize(value, unit);
-            }
+        Unit unit = Unit.fromUnitString(longOrDouble.group(2));
+        String number = longOrDouble.group(1);
+        if (number.indexOf('.') == -1) {
+            // Strings without decimals can avoid precision loss by parsing as long
+            return DataSize.of(Long.parseLong(number), unit);
+        }
+        else {
+            return new DataSize(roundDoubleSizeInUnitToLongBytes(Double.parseDouble(number), unit), unit);
         }
+    }
 
-        throw new IllegalArgumentException("Unknown unit: " + unitString);
+    private static long roundDoubleSizeInUnitToLongBytes(double size, Unit unit)
+    {
+        checkArgument(!Double.isInfinite(size), "size is infinite");
+        checkArgument(!Double.isNaN(size), "size is not a number");
+        checkArgument(size >= 0, "size is negative");
+        requireNonNull(unit, "unit is null");
+        double rounded = floor((size / (1.0d / unit.getFactor())) + 0.5d);
+        checkArgument(rounded <= Long.MAX_VALUE,
+                "size is too large to be represented in requested unit as a long");
```

Comment:
We might want to include the size+unit value here (same as in the `of()` factory)

Reply:
Added.